### DISCUSSION TITLE
Interop with JS and WASM from Rust Frameworks

### DISCUSSION THREAD

**Discussion by Qodestackr** [DioxusLabs/dioxus#3008] [General]
Hey everyone!

I’m currently working on a project where I want to integrate WASM into my web app for rendering radiology images. My setup involves using Remix for the shell, but for the complex, performance-heavy task of viewing DICOM images, I want to leverage Rust + WASM.

I love how Dioxus provides a React-like UI for Rust, but I’m curious if there are any best practices when using Dioxus with wasm-bindgen for seamless interop with JavaScript. Are there libraries, patterns, or specific tools I should be considering to make the integration easy to maintain?

My main goal is to build a performant radiology image viewer in WASM while keeping the maintenance as simple as possible.

Any guidance on this would be greatly appreciated!

Thanks in advance, Sorry if am abit vague, just thinking wasm-bindgen is too low leve for me 🙏

**Comment by ThomasCartier** [comment]
Hi,
You mean you already have an existing(and working)  JS stack ? Or you start from scratch and don't want to code your app in rust + very little JS glue?

if you can stay in the rust realm, do not hesitate, it's really easy with Dioxus.
otherwise, you will deal with [evals](https://dioxuslabs.com/learn/0.5/reference/web#running-javascript)

**Comment by Sitin** [comment]
It is not always possible to stay with Rust since this ecosystem is still missing important libraries. For example, try building an application with support of the custom maps and geo-data (JS ecosystem has plenty: [Leaflet](https://leafletjs.com/), [Maplibre](https://maplibre.org), [CesiumJS](https://cesium.com/platform/cesiumjs/), you name it). I hope this will change in the future (there are promissing early stage projects such as [galileo](https://github.com/Maximkaaa/galileo) or [maplibre-rs](https://github.com/maplibre/maplibre-rs)) but each developer has to pick their own fight and a good interop with JS is an important feature for frameworks like Dioxus.

At the moment, Dioxus `eval` provides a bare minimum of functionality. It would be nice to have something like [`tauri-specta`](https://github.com/specta-rs/tauri-specta) or another kind of typed interface between JavaScript and Dioxus.
